torch/csrc/jit/codegen/cuda/scheduler/matmul.cpp

```diff
@@ -40,6 +60,196 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
   tv->reorder(order_map);
 }
 
+//! Automatically generates the shared memory swizzled data layout
+//!  for matmul mainloop.
+//! The shared mem datalayout is always 2D currently, and this utility
+//!  function assumes that the innermost 2 dimensions on shared_mem_tv
+//!  are the ones begin swizzled.
+void prologSwizzle(TensorView* shared_mem_tv, const MatmulParam& params) {
+  // Check that the innermost 2 dimensions are concrete and static
+  //  sized so that the swizzle function can be defined.
+
+  // Utility to check concrete static size:
+  auto check_concrete_static_dim = [](IterDomain* id) {
+    TORCH_INTERNAL_ASSERT(
+        !id->isBroadcast() && !id->isReduction(),
+        "no support on reduction or broadcast dims");
+    TORCH_INTERNAL_ASSERT(
+        id->extent()->isConstInt(),
+        "swizzled dimensions need to be statically sized");
+  };
+
+  TORCH_INTERNAL_ASSERT(
+      shared_mem_tv->nDims() >= 2, "At least 2D input needed for swizzling");
+  check_concrete_static_dim(shared_mem_tv->axis(-2));
+  check_concrete_static_dim(shared_mem_tv->axis(-1));
+
+  auto mma_config = params.mma_builder.build();
+
+  // Extract the constant sizes of the swizzled tile
+  const auto tile_size_x = shared_mem_tv->axis(-2)->extent()->evaluateInt();
+  const auto tile_size_y = shared_mem_tv->axis(-1)->extent()->evaluateInt();
+
+  // TODO: add support for tf32(different macro) and fp32(ffma)
+  if (isTuring(mma_config.macro) || isAmpere(mma_config.macro)) {
```

Comment:
Having a hard time to get a clear idea of what this scheduling is going to do. Does any of the tests illustrate how this scheduling works?

Reply:
Yeah, this is one of the challenges in matmul kernels and I think the solution in this PR is quite interesting though convoluted. 

It is essentially solving the problem of `automatically generating a smem layout for a CTA tile for a mma type and bank conflict free` I'd probably need a separate docs and drawings to explain further into it.